### PR TITLE
[10.0][MIG] sale

### PR DESCRIPTION
**Pull request by rruebner**
FYI: I also updated the analysis text file for the sale module here. The rest of the analysis text files were updated in #850.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/openupgrade/doc/source/modules90-100.rst [hunk 1]
```diff
@@ -429,7 +429,7 @@ missing in the new release are marked with |del|.
 +-----------------------------------+-----------------------------------+
 |resource                           | Nothing to do                     |
 +-----------------------------------+-----------------------------------+
-|sale                               |                                   |
+|sale                               | Done                                  |
```
Comment (cubells) [inline odoo/openupgrade/doc/source/modules90-100.rst]: Typo: there are 4 extra spaces. :)

**Review by cubells — CHANGES_REQUESTED** [review]
Typo

File: addons/sale/migrations/10.0.1.0/openupgrade_analysis_work.txt [hunk 2]
```diff
@@ -14,7 +23,6 @@ sale         / sale.layout_category     / subtotal (boolean)            : NEW
 # NOTHING TO DO
```
Comment (pedrobaeza) [inline addons/sale/migrations/10.0.1.0/openupgrade_analysis_work.txt]: This is not correct, as sale now integrates sale_layout module. You need to run update_module_names with the merge option activated and check if there's any change on data model

File: addons/sale/migrations/10.0.1.0/openupgrade_analysis_work.txt [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
-# TODO: pre-migration: copy column
-# TODO: post-migration: map value 'cost' to new value
+# DONE: pre-migration: copy column
+# DONE: post-migration: map value 'cost' to new value
+
+sale         / product.template         / sale_line_warn (selection)    : NEW required: required, selection_keys: ['block', 'no-message', 'warning'], req_default: no-message
+sale         / product.template         / sale_line_warn_msg (text)     : NEW 
+sale         / res.partner              / sale_warn (selection)         : NEW required: required, selection_keys: ['block', 'no-message', 'warning'], req_default: no-message
+sale         / res.partner              / sale_warn_msg (text)          : NEW 
+# NOTHING TO DO
```
Comment (pedrobaeza) [inline addons/sale/migrations/10.0.1.0/openupgrade_analysis_work.txt]: Please check it if warning module was installed on v9 and rescue values from it.

File: addons/sale/migrations/10.0.1.0/post-migration.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+# © 2017 bloopark systems (<http://bloopark.de>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate(use_env=False)
+def migrate(cr, version):
+    # map old / non existing value 'cost' to default value 'order'
+    cr.execute("""
```
Comment (pedrobaeza) [inline addons/sale/migrations/10.0.1.0/post-migration.py]: You can use openupgradelib map_values function for readibility and extra log.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by rruebner** [comment]
@pedrobaeza I updated the migration parts according to your mentioned points. In addition I rechecked all parts again and added some more migration parts.

File: addons/sale/migrations/10.0.1.0/pre-migration.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        merge_modules=True
+    )
+
+
+def warning_update_module_names_partial(cr):
+    """We don't use openupgrade.update_module_names here because only the
+    fields sale_line_warn, sale_line_warn_msg, sale_warn and sale_warn_msg are
+    moved from the old warning module to the sale module. Other fields are
+    moved in other modules (e. g. field purchase_warn in purchase module). If
+    we would use openupgrade.update_module_names their might be problems when
```
Comment (pedrobaeza) [inline addons/sale/migrations/10.0.1.0/pre-migration.py]: s/their/there

File: addons/sale/migrations/10.0.1.0/pre-migration.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    first use openupgrade.update_module_names in sale module and than again in
```
Comment (pedrobaeza) [inline addons/sale/migrations/10.0.1.0/pre-migration.py]: s/use/using

Comment (pedrobaeza) [reply addons/sale/migrations/10.0.1.0/pre-migration.py]: s/than/then

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Check if you need to rename the security groups

**Comment by rruebner** [comment]
@pedrobaeza I fixed the mentioned typos and rechecked security group changes. AFAICS there is no extra renaming needed (possible renaming is covered by update_module_names calls).
I also checked for noupdate=1 records and everything was fine.

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM! :+1:

**Comment by pedrobaeza** [comment]
Thanks for the changes, Robert.

If you don't mind, I'm going to keep the PRs unmerged until we get a green branch for not adding more complexity to the pipe and possible side effects.

**Comment by MiquelRForgeFlow** [comment]
Please, rebase the PR (and squash minor commits) ☺️

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]
A new line appeared https://github.com/OCA/OpenUpgrade/pull/917/files#diff-1c07737aeb217d5e147f8fa6d1204414R5, and it should be attended in an easy post-migration script.

File: addons/sale/migrations/10.0.1.0/pre-migration.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+# © 2017 bloopark systems (<http://bloopark.de>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
+
+
+# backup of separator column in table sale_layout_category because it doesn't
+# exist anymore in odoo 10
+column_renames = {
```
Comment (omar7r) [inline addons/sale/migrations/10.0.1.0/pre-migration.py]: If you don't have installed "sale_layout" module, columns renaming and the models renaming fails, maybe, you have to check if this module is installed to execute this two functions

**Comment by pedrobaeza** [comment]
This fails due to TRAVIS_COMMIT, but I have checked it and fix it, so I merge.

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**